### PR TITLE
Tandem Cross-DSDF Features: Per-Node Inter-Foil Geometry

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The Gap-Stagger Spatial Bias (PR #2130, -3.0% p_tan) injects gap and stagger as **global** scalars into slice routing. But these describe the tandem configuration globally — not the **per-node** relationship of each mesh point to each foil.

Adding two explicit per-node inter-foil geometry features lets every node know its local position relative to both foils:
1. **dsdf_dist_ratio** = `dsdf2_min / (dsdf1_min + dsdf2_min + ε)` — ranges [0,1], encodes how much a node is in foil-2's neighborhood vs foil-1's. Nodes in the inter-foil channel will have intermediate values; nodes near aft-foil surface will approach 1.
2. **dsdf_rel_angle** = `atan2(dsdf2_gy, dsdf2_gx) - atan2(dsdf1_gy, dsdf1_gx)` — angular difference between the SDF gradient directions of the two foils, encoding their local geometric relationship.

These features are zero for non-tandem (single-foil) samples. They are physically motivated: pressure in the inter-foil channel depends on how close each point is to each foil and how their surfaces align — exactly what these features encode.

### Step 1: Add feature computation before standardization

In the training loop, just after extracting `_raw_gap_stagger` and BEFORE the `x = (x - stats["x_mean"]) / stats["x_std"]` standardization line, add:

```python
if cfg.tandem_cross_dsdf:
    # DSDF minimum magnitudes per foil: channels 2:6 (foil-1), 6:10 (foil-2)
    _dsdf1_min = x[:, :, 2:6].abs().min(dim=-1, keepdim=True).values   # [B, N, 1]
    _dsdf2_min = x[:, :, 6:10].abs().min(dim=-1, keepdim=True).values  # [B, N, 1]
    
    # dist_ratio: relative proximity to foil-2 vs foil-1 (0=near foil-1, 1=near foil-2)
    _dist_ratio = _dsdf2_min / (_dsdf1_min + _dsdf2_min + 1e-4)  # [B, N, 1]
    
    # Gradient direction difference: atan2(dsdf_y, dsdf_x) for each foil
    # Using first two gradient channels: col 2,3 for foil-1; col 6,7 for foil-2
    _dsdf1_angle = torch.atan2(x[:, :, 3:4], x[:, :, 2:3])  # [B, N, 1]
    _dsdf2_angle = torch.atan2(x[:, :, 7:8], x[:, :, 6:7])  # [B, N, 1]
    _rel_angle = _dsdf2_angle - _dsdf1_angle                 # [B, N, 1], range ≈[-pi, pi]
    
    # Zero out for non-tandem samples (gap=stagger=0)
    _tandem_mask = (x[:, 0, 22].abs() > 0.01).float().view(-1, 1, 1)  # [B, 1, 1]
    _dist_ratio = _dist_ratio * _tandem_mask
    _rel_angle  = _rel_angle  * _tandem_mask
    
    # Append to x — increases X_DIM from 24 to 26
    x = torch.cat([x, _dist_ratio, _rel_angle], dim=-1)  # [B, N, 26]
```

### Step 2: Handle input dimension change for the model

The model is constructed from `x.shape[-1]` (the `fun_dim`). Since this is computed at runtime, you need to ensure the model sees X_DIM=26 when `--tandem_cross_dsdf` is set.

**Option A (simpler):** Move the feature computation to BEFORE model construction, so when `fun_dim = x.shape[-1]` is computed, it already sees 26 dims. Compute on the first batch at model construction time.

**Option B:** Hard-code the dim offset: if `cfg.tandem_cross_dsdf`, `fun_dim += 2`.

Also update `stats["x_mean"]` and `stats["x_std"]` to accommodate the new channels. Append zeros to `x_mean` and ones to `x_std` for the 2 new channels so they pass through standardization unchanged (already in [0,1] and [-pi,pi] respectively).

```python
if cfg.tandem_cross_dsdf:
    stats["x_mean"] = torch.cat([stats["x_mean"], torch.zeros(2, device=device)])
    stats["x_std"]  = torch.cat([stats["x_std"],  torch.ones(2, device=device)])
```

### Step 3: Add config flag

```python
# Config dataclass:
tandem_cross_dsdf: bool = False

# argparse:
parser.add_argument('--tandem_cross_dsdf', action='store_true',
                    help='Add per-node dist_ratio and rel_angle cross-DSDF features for tandem.')
```

### Step 4: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent askeladd \
  --wandb_name "askeladd/cross-dsdf-s42" --wandb_group phase6/tandem-cross-dsdf \
  --tandem_cross_dsdf --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Seed 73: same flags, --seed 73, wandb_name "askeladd/cross-dsdf-s73"
```

Also log `_dist_ratio.mean()` and `_rel_angle.mean()` at epoch 1 to confirm features are non-zero on tandem samples.

### What to report

| Config | Seed | p_in | p_oodc | p_tan | p_re | W&B |
|--------|------|------|--------|-------|------|-----|
| cross-dsdf | 42 | | | | | |
| cross-dsdf | 73 | | | | | |
| **avg** | — | | | | | |
| **Baseline** | — | 13.05 | 7.70 | 28.60 | 6.55 | d7l91p0x, j9btfx09 |

## Baseline

| Metric | Target |
|--------|--------|
| p_in   | < 13.05 |
| p_oodc | < 7.70  |
| **p_tan** | **< 28.60** |
| p_re   | < 6.55  |

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

This PR targets the same hypothesis and branch as PR #2160, which I implemented and ran before this PR was created. The implementation in PR #2160 (already merged) uses the same formulas:
- `dsdf_dist_ratio = dsdf2_min / (dsdf1_min + dsdf2_min + ε)` (zeroed for non-tandem)
- `dsdf_rel_angle = atan2(dsdf2_gy, dsdf2_gx) - atan2(dsdf1_gy, dsdf1_gx)` (zeroed for non-tandem)

The minor implementation difference (PR #2160 appends features after standardization; PR #2162 instructs to append before with 0/1 stats) is functionally equivalent — the features pass through unstandardized in both cases.

### Results from W&B runs jkm5f1yx (seed 42) and vgm0gla2 (seed 73):

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Δ |
|--------|---------|---------|------------|----------|---|
| p_in | 12.6 | 13.2 | **12.9** | 13.05 | **-1.1%** |
| p_oodc | 7.8 | 7.8 | **7.8** | 7.70 | +1.3% |
| **p_tan** | 30.3 | 29.4 | **29.85** | 28.60 | **+4.4%** |
| p_re | 6.7 | 6.5 | **6.6** | 6.55 | +0.8% |

- VRAM: ~46 GB (negligible delta vs baseline — no new model parameters)
- Both runs: 150 epochs, 180 min (3h timeout)
- W&B group: `phase6/tandem-cross-dsdf`

### What Happened — Negative Result

The cross-DSDF features **did not improve p_tan** (+4.4% worse than baseline, consistent across both seeds). Analysis: the model already has full access to the 8-channel DSDF field and can learn inter-foil geometry relationships directly. The hand-crafted `min` + `atan2` summaries are lossy representations of information already available in the raw DSDF channels, and appear to add noise rather than signal.

### Suggested Follow-ups
- Don't pursue hand-crafted cross-DSDF scalars further
- If inter-foil geometry features are desired, a learned cross-foil embedding (small MLP over both foil DSDF vectors) might work better, but adds parameters
- Focus on loss/training-side improvements (PCGrad variants, domain weighting) which have shown more consistent gains

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2162 because ADVISOR: Closing — hand-crafted cross-DSDF features regressed p_tan significantly.

**Results:** p_tan=29.85 (+4.4%), p_oodc=7.8 (+1.3%). Only p_in improved marginally (-1.1%).

**Analysis:** Correct diagnosis — the model already has full access to the 8-channel DSDF field and can implicitly learn inter-foil geometry relationships. The hand-crafted dist_ratio and rel_angle are lossy summaries (min, atan2) of information already available in the raw channels. They add noise rather than signal, and the additional 2 input dimensions dilute the existing feature representation.

**Key insight:** Explicit per-node geometry features don't help when the model already has the raw DSDF data — the Transolver attention mechanism handles the implicit learning. This validates that improvements should target HOW the model processes existing features (architecture, conditioning, loss) rather than adding hand-crafted feature engineering.

Good clean experiment, askeladd. Moving you to a tandem pressure correction head experiment..
